### PR TITLE
Write Functionality

### PR DESCRIPTION
**Pull request by Billcountry**
Introduces the following functions for writing:
- put, put_async, delete, delete_async

**Behaviour Notes**
- Put on datastore only happens only when there was a change in the data
- This also applies to the `post_put` function

### CODE REVIEW

File: tests/test_entity.py [hunk 1]
```diff
@@ -77,4 +88,108 @@ def test_get(self):
...
+        self.assertNotEqual(saved_entity, entity)
+        entity.prop1 = "Text values"
+        self.assertEqual(saved_entity, entity)
+        # Compare entities with different keys but equal values
+        separate_entity = TestEntity(required_property=entity.required_property,
+                                     required_default=entity.required_default, prop1=entity.prop1, prop2=entity.prop2)
+        self.assertNotEqual(entity, separate_entity)
+        separate_entity.put()
+        self.assertNotEqual(saved_entity, separate_entity)
+        self.assertEqual(saved_entity, entity)
```
Comment (georgemarrows) [inline tests/test_entity.py:110]: Isn't this the same as line 103? It's a bit confusing here
```suggestion
```

File: echo/datastore/db_utils.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+def put_async(entities):
+    """Save to datastore asynchronously
+    Args:
+        entities: An entity or a list of entities
+    Returns:
+        A future object. Use future.result() to wait for or get the results if ready
+    """
+    if not hasattr(entities, "__iter__"):
+        entities = [entities]
+    with ThreadPoolExecutor(max_workers=2) as executor:
```
Comment (georgemarrows) [inline echo/datastore/db_utils.py:24]: Doesn't the library we're wrapping provide async puts, deletes etc?

Comment (georgemarrows) [inline echo/datastore/db_utils.py:24]: Should we create a `ThreadPoolExecutor` on each call to `put_async` like this? 

I haven't checked the docs (doh!) but if it behaves similarly to comparable Java classes, you'd normally create a thread pool once and use it again and again in the application.

File: echo/datastore/properties.py [hunk 3]
```diff
@@ -76,12 +81,19 @@ def user_value(self, value):
...
+    """
     def __init__(self, auto_now_add=False, required=False):
         default = datetime.now() if auto_now_add else None
         super(DateTimeProperty, self).__init__(default=default, required=required)
 
     def validate(self, user_value):
-        return self.__type_check__(user_value, datetime)
+        user_value = self.__type_check__(user_value, datetime)
+        if user_value and not user_value.tzinfo:
+            user_value = pytz.utc.localize(user_value)
```
Comment (georgemarrows) [inline echo/datastore/properties.py:95]: I don't think I saw a test for this. It's important.

File: echo/datastore/entity.py [hunk 4]
```diff
@@ -1,11 +1,14 @@
 from google.cloud.datastore import Client, Entity as DatastoreEntity, Key as DatastoreKey
 from six import string_types
-from future import builtins
-from echo.datastore.errors import InvalidKeyError, NotSavedException
-from echo.datastore import properties
+from echo.datastore.errors import InvalidKeyError, NotSavedException, InvalidValueError
+from echo.datastore import properties, db_utils
+
+__pdoc__ = {}
```
Comment (georgemarrows) [inline echo/datastore/entity.py:6]: What's this for?

File: echo/datastore/entity.py [hunk 5]
```diff
@@ -21,41 +24,51 @@ class Entity(object):
...
 
     def __init__(self, **data):
         if type(self) is Entity:
             raise Exception("You must extend Entity")
-        self.__id = None
-        if "id" in data:
-            self.__id = data.get("id")
-        self.__datastore_entity__ = DatastoreEntity(key=self.key(partial=True))
+        project = db_utils.__client__().project
+        self.__changes__ = []
```
Comment (georgemarrows) [inline echo/datastore/entity.py:36]: What do we track `__changes__` for in such detail?

File: tests/test_propterty.py [hunk 6]
```diff
@@ -28,8 +29,6 @@ def test_basic_property(self):
 
         # Default value should be set on get
         self.assertEqual(self.entity.integer_property_with_default, 10)
-        # You can't set a required property to null
-        self.assertRaises(errors.InvalidValueError, setattr, self.entity, "integer_property_required", None)
```
[the remark this answers is not in the document]
Comment (georgemarrows) [reply tests/test_propterty.py:32]: I think it just moved somewhere else didn't it?

File: requirements.txt [hunk 7]
```diff
@@ -1,2 +1,3 @@
 google-cloud-datastore==1.11.0
 future==0.18.2
+futures==3.1.1
```
[the remark this answers is not in the document]
Comment (Billcountry) [reply requirements.txt:3]: - `future` allows compatibility between python2 and 3. e.g. getting builtins in both versions
- `futures` is used to run asynchronous tasks, provides thread pool executor

I've added both as comments on the code

File: tests/test_propterty.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_propterty.py:32]: It was in a way changed. While a required property must be set, this shouldn't mean that the value can't be null.
We will throw an error if  the value is not set at all, however it's acceptable to explicitly set a null

File: echo/datastore/db_utils.py (hunk 2 above)
Comment (Billcountry) [reply echo/datastore/db_utils.py:24]: It doesn't.

File: echo/datastore/entity.py (hunk 4 above)
Comment (Billcountry) [reply echo/datastore/entity.py:6]: It's used by pydoc to override methods and classes that we don't wan to expose. e.g. `BaseEntityMeta`
Adding a comment on the code

File: echo/datastore/db_utils.py (hunk 2 above)
Comment (Billcountry) [reply echo/datastore/db_utils.py:24]: Done, created once and stored in builtins

File: echo/datastore/entity.py (hunk 5 above)
Comment (Billcountry) [reply echo/datastore/entity.py:36]: Looking at the problems we face internally, tracking changes would solve a lot of problems for us, e.g. Saving hashes of data to check if it has changed and make an update to a different subsystem.

Comment (georgemarrows) [reply echo/datastore/entity.py:36]: Sounds useful. Let's see how it performs in terms of memory usage in the app.

File: echo/datastore/db_utils.py (hunk 2 above)
Comment (georgemarrows) [reply echo/datastore/db_utils.py:24]: Now something else occurs to me. By using threads to execute requests like this, we're assuming that the new library is thread-safe, and can cope with multiple threads running through it at the same time. That's not all guaranteed. Do their docs say if it is / isn't? If not, would be worth a question to the team that develops it.

File: echo/datastore/properties.py (hunk 3 above)
Comment (Billcountry) [reply echo/datastore/properties.py:95]: Actually tests fail on python3.8 because of this. Equality on python3.7 seems to ignore timezone info

Comment (Billcountry) [reply echo/datastore/properties.py:95]: Tested

File: tests/test_entity.py (hunk 1 above)
Comment (Billcountry) [reply tests/test_entity.py:110]: Thanks, removing.

File: echo/datastore/db_utils.py (hunk 2 above)
Comment (Billcountry) [reply echo/datastore/db_utils.py]: The anwer moved from 2017
> [The short answer is: We think so. :-)](https://github.com/googleapis/google-cloud-python/issues/3272#issuecomment-291874468)

Down the line on the same thread we get:
> [We expect that gRPC-based clients to be thread safe: the issues we know of are to do with multiprocessing](https://github.com/googleapis/google-cloud-python/issues/3272#issuecomment-492405446)

I'll write a hack to test this on the python3 service

Comment (georgemarrows) [reply echo/datastore/db_utils.py]: Thanks for investigating and doing our own tests.

You're now Github-famous
image

Comment (Billcountry) [reply echo/datastore/db_utils.py]: @georgemarrows running the code below directly on app engine works as expected, creates and deletes the entities
```py
from echo.datastore import db, Entity
from api.handlers import RouteHandler
import logging


class Thread(Entity):
    name = db.TextProperty()
    number = db.IntegerProperty()
    created = db.DateTimeProperty(auto_now_add=True)


def threads(handler: RouteHandler):
    logging.debug("Creating permanent threads")
    permanent = [Thread(name="Permanent Thread %s" % i, number=i) for i in range(1, 129)]
    logging.debug("Creating temporary threads")
    temp = [Thread(name="Permanent Thread %s" % i, number=i) for i in range(129, 257)]
    logging.debug("Saving threads")
    db.put(temp)
    db.put(permanent)
    logging.debug("Deleting temporary threads")
    db.delete(temp)
    return [{"key": thread.key().id_or_name, "name": thread.name} for thread in permanent]
```

**Pull request merged**